src/gc.c

```diff
@@ -2277,10 +2435,51 @@ JL_DLLEXPORT void jl_gc_collect(int full)
 #endif
     if (recollect) {
         n_pause--;
-        jl_gc_collect(0);
+        _jl_gc_collect(0, stack_hi);
     }
 }
 
+JL_DLLEXPORT void jl_gc_collect(int full)
+{
+    if (!is_gc_enabled || jl_in_gc)
+        return;
```

Comment:
i didn't realize jl_in_gc was thread-local now. it seems like that flag should never be true here now in that case?

to return to the original purpose of this flag, it should perhaps be renamed `jl_in_finalizers` and moved to only protecting the finalizer queue from task switches? the original intent of this flag was to disallow finalizers from descheduling the thread and switching to running other tasks at unexpected yield points (any allocation).